### DISCUSSION TITLE
Can we invoke Client side Function based on LLMs/Chains decision to run one?

### DISCUSSION THREAD

**Discussion by AndriiTsok** [langchain-ai/langserve#334] [Q&A] [answered]
Can we perform Client side Functions based on LLMs/Chains decision to run one? May be there is a way to request/and supply values for function calling from client to server?

**Comment by eyurtsev** [comment, marked as answer]
What format is the llm / chain returning for a function invocation request?

**Reply by AndriiTsok** [reply]
Hi @eyurtsev We're developing a chat assistant leveraging LangServe. Our system primarily uses OpenAI for executing functions and managing the assistant's persona. While we have successfully integrated server-side function execution through LangChain, we occasionally face situations where we need to call a function that is implemented on the client side. 

I am wondering if LangServe and JS client support the capability for LangServe to direct the client to perform specific function calls during an ongoing conversation? 

For instance, one of the functions can be: `requestCustomerContactInformation` with parameters [name, email, phone]. On the client this function would display a form to fill in the fields. By filling in the form we would provide the LLM (through LangServe) the values of the function call.

Thank you for all your work and amazing ecosystem of the libraries! If this is not supported by LangServe/JS client. I would love to contribute.

**Reply by eyurtsev** [reply]
You can probably achieve that by routing the tool invocation requests from the assistant to either the server or the client based on where the tool is available (server or client).

Anything in this list of messages will serialize: https://github.com/langchain-ai/langserve/blob/main/langserve/serialization.py#L63

And python RemoteRunnable client should be able to de-serialize all of those messages (js client too though it may not be 100% up to date).

If the tool is on the client side, the server returns a function invocation request, and the client executes the tool and returns a tool result. The client may need to include conversation history if conversation history is tracked on the front end.

The tricky part is the routing of the message using the existing agent executor abstraction in langchain (if you're using it).

I suspect you could achieve this by parameterizing the parser that the agent is using, such that it yields an `AgentFinish` for any tool invocation request that is defined on the client side. You can attach to the agent finish the actual AIMessage (in return_values).
